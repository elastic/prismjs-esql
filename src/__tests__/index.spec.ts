import {test, expect} from 'vitest';
import {refractor} from 'refractor';
import type {Syntax, RefractorRoot, RefractorElement} from 'refractor';
import {esql} from '..';

refractor.register(esql as Syntax);

type Token = [text: string, type: string];

const walk = (prefix: string, node: RefractorRoot | RefractorElement): Token[] => {
  const tokens: Token[] = [];

  for (const child of node.children) {
    if (child.type === 'element') {
      const classNames = child.properties?.className
        ? Array.isArray(child.properties?.className)
          ? child.properties.className.join('+')
          : String(child.properties?.className)
        : '';
      const type = prefix ? prefix + '.' + classNames : classNames;

      tokens.push(...walk(type, child));
    }
    if (child.type === 'text') {
      tokens.push([child.value, prefix]);
    }
  }

  return tokens;
};

const tokenize = (src: string): Token[] => {
  const root: RefractorRoot = refractor.highlight(src, 'esql');
  const tokens: Token[] = walk('', root);

  return tokens;
};

test('basic FROM command', () => {
  const tokens = tokenize('FROM index');

  expect(tokens).toMatchSnapshot();
});

test('comments', () => {
  const tokens = tokenize(`// This is comment
/* This is multiline comment */`);

  expect(tokens).toMatchSnapshot();
});

test('booleans', () => {
  const tokens = tokenize(`ROW TRUE, FALSE`);

  expect(tokens).toMatchSnapshot();
});

test('numbers', () => {
  const tokens = tokenize(`ROW 1, 3.14`);

  expect(tokens).toMatchSnapshot();
});

test('strings', () => {
  const tokens = tokenize(`ROW "abc"`);

  expect(tokens).toMatchSnapshot();
});

test('functions', () => {
  const tokens = tokenize(`ROW AVG(COUNT(123))`);

  expect(tokens).toMatchSnapshot();
});

test('cast', () => {
  const tokens = tokenize(`ROW 1.1::INTEGER`);

  expect(tokens).toMatchSnapshot();
});
