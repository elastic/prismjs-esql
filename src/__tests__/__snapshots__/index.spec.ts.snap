// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`basic FROM command 1`] = `
[
  [
    "FROM",
    "token+command+keyword",
  ],
  [
    " index",
    "",
  ],
]
`;

exports[`booleans 1`] = `
[
  [
    "ROW",
    "token+command+keyword",
  ],
  [
    " ",
    "",
  ],
  [
    "TRUE",
    "token+boolean",
  ],
  [
    ",",
    "token+punctuation",
  ],
  [
    " ",
    "",
  ],
  [
    "FALSE",
    "token+boolean",
  ],
]
`;

exports[`cast 1`] = `
[
  [
    "ROW",
    "token+command+keyword",
  ],
  [
    " ",
    "",
  ],
  [
    "1.1",
    "token+float+number",
  ],
  [
    "::INTEGER",
    "token+cast+operator",
  ],
]
`;

exports[`comments 1`] = `
[
  [
    "// This is comment",
    "token+comment",
  ],
  [
    "
",
    "",
  ],
  [
    "/* This is multiline comment */",
    "token+multiline-comment+comment",
  ],
]
`;

exports[`functions 1`] = `
[
  [
    "ROW",
    "token+command+keyword",
  ],
  [
    " ",
    "",
  ],
  [
    "AVG",
    "token+function",
  ],
  [
    "(",
    "token+punctuation",
  ],
  [
    "COUNT",
    "token+function",
  ],
  [
    "(",
    "token+punctuation",
  ],
  [
    "123",
    "token+integer+number",
  ],
  [
    ")",
    "token+punctuation",
  ],
  [
    ")",
    "token+punctuation",
  ],
]
`;

exports[`numbers 1`] = `
[
  [
    "ROW",
    "token+command+keyword",
  ],
  [
    " ",
    "",
  ],
  [
    "1",
    "token+integer+number",
  ],
  [
    ",",
    "token+punctuation",
  ],
  [
    " ",
    "",
  ],
  [
    "3.14",
    "token+float+number",
  ],
]
`;

exports[`strings 1`] = `
[
  [
    "ROW",
    "token+command+keyword",
  ],
  [
    " ",
    "",
  ],
  [
    ""abc"",
    "token+string",
  ],
]
`;
